### PR TITLE
[benchmark] Janitor Duty: Sweep Quadratic

### PR DESCRIPTION
**Pull request by palimondo**
To enable [robust performance measurements](https://forums.swift.org/t/towards-robust-performance-measurement/11490) by minimizing the accumulated error, this splits the composite tests from `DictionatyCopy` and `DictionaryFilter` into smaller individual benchmarks by dictionary size. These benchmarks guard against the quadratic behavior ([SR-3268](https://bugs.swift.org/browse/SR-3268)). Also removes the older disabled `HashQuadratic` benchmark that was covering the same issue.

quadratictests

### CODE REVIEW

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci please benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci benchmark

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please smoke test

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@swift-ci please test

**Comment by palimondo** [comment]
@eeckstein @lorentey Please review 🙏

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@swift-ci smoke test linux platform

**Comment by palimondo** [comment]
@eeckstein I believe this is now mergeable: the full test that runs `Benchmark_O.test.md` passed on mac os and the previously broken Linux smoke test (unrelated to this PR) now also passed (full test there is not required as the benchmark validation does not run on Linux).

**Comment by eeckstein** [comment]
I'll let @lorentey review this

File: benchmark/single-source/DictionaryCopy.swift [hunk 1]
```diff
@@ -17,66 +17,54 @@
 
 import TestsUtils
 
-public let DictionaryCopy = [
-  BenchmarkInfo(name: "DictionaryCopy", runFunction: run_DictionaryCopy,
-    tags: [.validation, .api, .Dictionary]),
-  BenchmarkInfo(name: "DictionaryFilter", runFunction: run_DictionaryFilter,
-    tags: [.validation, .api, .Dictionary]),
-]
+let t: [BenchmarkCategory] = [.validation, .api, .Dictionary]
```
Comment (lorentey) [inline benchmark/single-source/DictionaryCopy.swift]: Nit: Please expand this name to `tags` or somesuch -- we generally dislike single-character names in the codebase.

**Review by lorentey — CHANGES_REQUESTED** [review]
The code reorganization looks good to me! I don't much mind losing the old variants here -- although I guess they did have some value in release-to-release comparisons, I expect we have other benchmarks for insertion at least.

I have one question about the results, though: If I read the benchmark results correctly, the Quadratic Surprise doesn't show up in the filter benchmarks. I can see the bump in `Dict.CopyKeyValue.20k`, but there is no corresponding increase for `Dict.FilterAllMatch.20k` -- all the filter results seem to be roughly the same.

The implementation of filter will change at some point soon, but I'm wondering if 16k--28k is large enough to trigger the issue.

File: benchmark/single-source/DictionaryCopy.swift (hunk 1 above)
Comment (palimondo) [reply benchmark/single-source/DictionaryCopy.swift]: I have been using `t` for reducing this useless boilerplate throughout many of my past Janitor Duty PRs and Erik never objected… Do you insist on this change?

I'm also planning to propose removing the tags altogether and keeping only the `skip`, as it seems like an over-engineered aspect of out benchmark suite. Do you know if anybody uses these?

**Comment by palimondo** [comment]
I have also noticed the filter has been much flatter in response, but the quadratic behavior is still fully visible in the `-Onone `build. The `filter` in optimized builds had less pronounced peak at the low end since PR #19213 when you switched it to work directly with [native dictionary](https://github.com/apple/swift/pull/19213/files#diff-74d864a8dd57b951efde782cf2240fc5R605).

File: benchmark/single-source/DictionaryCopy.swift (hunk 1 above)
Comment (palimondo) [reply benchmark/single-source/DictionaryCopy.swift]: @lorentey Does the above argument persuade you? Or is it a blocker?

Comment (lorentey) [reply benchmark/single-source/DictionaryCopy.swift]: It's a nitpick. I didn't know this is a pattern now -- we can fix it later.

**Comment by lorentey** [comment]
Ah, that makes sense 👍

**Review by lorentey — APPROVED** [review]

**Comment by palimondo** [comment]
@lorentey Thank you!

**Pull request merged**